bpf/lib/arp.h

```diff
@@ -19,13 +19,15 @@ struct arp_eth {
 
 /* Check if packet is ARP request for IP */
 static __always_inline int arp_check(struct ethhdr *eth, struct arphdr *arp,
-				     union macaddr *mac)
+				     struct arp_eth *arp_eth,
+				     union macaddr *mac, __be32 *ip)
 {
 	union macaddr *dmac = (union macaddr *) &eth->h_dest;
 
 	return arp->ar_op  == bpf_htons(ARPOP_REQUEST) &&
 	       arp->ar_hrd == bpf_htons(ARPHRD_ETHER) &&
-	       (eth_is_bcast(dmac) || !eth_addrcmp(dmac, mac));
+	       (eth_is_bcast(dmac) || !eth_addrcmp(dmac, mac)) &&
+	       (ip && *ip == arp_eth->ar_tip);
```

Comment:
My other thought that I was wondering about is basically, I've considered Cilium to be sort of an 'l3 device' from the perspective of the endpoint network, terminating the L2 domain at the host-side of the veth pair*. While we do currently have the code to pass unknown ARP requests to the Linux stack, I don't really know what kind of requests would do so and it's not clear to me that this is either (a) correct or (b) will do the right thing (eg for forwarding on the reverse path). I don't actually know of any other particular ARP message types that might fall into this category.

I can certainly see the argument that we should not respond to ARPs for the endpoint's IP. But then Cilium is typically responsible for configuring that IP*, so why is the endpoint performing duplicate address detection in the first place? Could we just disable that functionality? Then there's the case of other IPs. Cilium makes no guarantee of routability of IPs that it hasn't assigned to the endpoint, so while it's not strictly correct for Cilium to respond to ARPs for any other random IP, even if we avoid responding, the connectivity probably won't work anyway. There's a certain beauty in the simplicity of not having to track anything else like gateway IPs.

[*] Modulo chaining modes, where I seem to recall we have some logic to disable the ARP responder and support passthrough instead.

Thoughts? Background? Any context I'm missing here?

Reply:
(I also note that in the past, we also avoided tracking gateway IPs because the compile-time vs. load-time configuration stuff was not implemented, so functionality like what's proposed here was too costly. Now it should be practically free from an endpoint datapath configuration perspective, though does affect complexity / debuggability of the datapath).

Reply:
> My other thought that I was wondering about is basically, I've considered Cilium to be sort of an 'l3 device' from the perspective of the endpoint network, terminating the L2 domain at the host-side of the veth pair*. While we do currently have the code to pass unknown ARP requests to the Linux stack, I don't really know what kind of requests would do so and it's not clear to me that this is either (a) correct or (b) will do the right thing (eg for forwarding on the reverse path). I don't actually know of any other particular ARP message types that might fall into this category.

I am not really sure either, but in this respect cilium would be behaving as if it was not really there and I think that is the safe approach.

> I can certainly see the argument that we should not respond to ARPs for the endpoint's IP. But then Cilium is typically responsible for configuring that IP*, so why is the endpoint performing duplicate address detection in the first place? Could we just disable that functionality? Then there's the case of other IPs. Cilium makes no guarantee of routability of IPs that it hasn't assigned to the endpoint, so while it's not strictly correct for Cilium to respond to ARPs for any other random IP, even if we avoid responding, the connectivity probably won't work anyway. There's a certain beauty in the simplicity of not having to track anything else like gateway IPs.
> 
> [*] Modulo chaining modes, where I seem to recall we have some logic to disable the ARP responder and support passthrough instead.
> 
> Thoughts? Background? Any context I'm missing here?

This particular scenario I am attempting to solve is kubevirt bridged setup, where the endpoint side interface is placed on a bridge together with a VM tap, the ip address is unassigned from it and re-assigned to the VM interface. This VM could be running any image and is typical from some systems to do this IP address duplication check on assignment.

Also with things like cni chaining, multus... involving a variety of plugins around, you never know what else is going on. So while every other plugin would have to ensure it's own things that it needs, for the resulting integration to work plugins probably expect some nominal behavior from each other. 

So it all drives the point again of taking the safest approach of behaving as it would be expected in all fronts possible?

Reply:
> This particular scenario I am attempting to solve is kubevirt bridged setup, where the endpoint side interface is placed on a bridge together with a VM tap, the ip address is unassigned from it and re-assigned to the VM interface. 

This sounds effectively like a chaining mode: Cilium is not responsible for base connectivity since the VM taps are being created and associated with a bridge. This logic is all defined outside of Cilium so I'd expect Cilium to be configured to explicitly avoid handling L2.

> Also with things like cni chaining, multus... involving a variety of plugins around, you never know what else is going on.

I see there being two clear categories: Either Cilium is responsible for L2 connectivity, or Cilium is delegating L2 connectivity to another tool. We know which of these paths we are taking, and this maps down to `#define ENABLE_ARP_RESPONDER` and/or `#define ENABLE_ARP_PASSTHROUGH` in the datapath side.

Do you explicitly need Cilium to handle some ARP in your scenario? Can we avoid tracking the gateway IPs by just configuring Cilium to enable passthrough instead?

Reply:
> Do you explicitly need Cilium to handle some ARP in your scenario? Can we avoid tracking the gateway IPs by just configuring Cilium to enable passthrough instead?

I don't see a way to generally make use of the ARP passthrough option but I might be missing something. Wouldn't using this option bring back the possible problem of endpoints with old gateways? Isn't this option primarily thought to be used by cilium when cilium is not the main CNI? Would other tools be required to be specifically aware of cilium and this option to exercise it? Or are we talking about globally and permanently enable this option on deployment? Isn't it a per endpoint setting?

> This sounds effectively like a chaining mode: Cilium is not responsible for base connectivity since the VM taps are being created and associated with a bridge. This logic is all defined outside of Cilium so I'd expect Cilium to be configured to explicitly avoid handling L2.

Just to make sure, in this case and other cases I am referring to, cilium is the main CNI and other secondary CNIs either in chain scenarios or multus type meta CNIs are the ones doing extra stuff on an assumed or expected behavior of the primary CNI. In this particular case, however, not even that. Just setting up a bridge in the pod with additional interfaces and moving the IP around in a way that should no affect the underlying service provided by the CNI.

Isn't it at the end a question of improving cilium's L2 handling in this matter vs looking for a workaround somewhere else?

Reply:
> Isn't this option primarily thought to be used by cilium when cilium is not the main CNI?

Yes.

> In this case and other cases I am referring to, cilium is the main CNI ...

OK cool, so the answer to `Do you explicitly need Cilium to handle some ARP in your scenario?` is "yes"; Cilium is handling L2.

> In this particular case, however, not even that. Just setting up a bridge in the pod with additional interfaces and moving the IP around in a way that should no affect the underlying service provided by the CNI.

I agree, I misread the previous comment which lead me to think that other CNIs were configuring the host-side interfaces. Sounds like this isn't the case.